### ISSUE TITLE
[Refactor] Let Reef own inference independently of training backends

### ISSUE THREAD

**Issue by BobbyZhouZijian** [Human-Agent-Society/reef#425]
### Primary area

Runtime & Serving; training backends

### Summary and decision to make

**Proposed:** Reef owns the coordination of independent inference and training backends. `reef/runtime/` contains Reef's contracts and scheduling logic. Concrete inference implementations live in `reef/inference/<backend>/`; training implementations live in `reef/train/<backend>/`. Neither integration imports the other.

Reef owns component startup and shutdown, resource handoff, weight-version assignment and verification, publication, commit acknowledgement and recovery. Backends implement native model operations and compatible weight transport. `TrainingRuntime` and `InferenceRuntime` are independent interfaces; no aggregate `ModelRuntime` is introduced.

This follows #414/#424. [Implementation PR #432](https://github.com/Human-Agent-Society/reef/pull/432) remains a draft. RFC status is **proposed**: maintainer acceptance and GPU validation remain open. This issue is not closed by the implementation PR.

### Motivation and supporting results

At #424 commit `81cc9c9`, Slime's bootstrap created placement groups, its rollout manager, inference engines and training groups. Moving resource flags into `inference.*` improved configuration ownership but did not change lifecycle ownership. Later extraction exposed separate runtime objects, yet Slime's bridge still implemented publication, LoRA residency and training/inference phase ordering.

Future inference and training backends need the same coordination rules without depending on Slime. The shared scheduler should decide when a version may serve, while backend adapters implement how to train, transfer or receive it. These findings come from code inspection; no GPU performance result is claimed.

### Goals and non-goals

Goals:

- Independently select inference and training implementations, validating their transport compatibility before allocation.
- Keep public inference settings under `inference.*` and training settings under `training.*`, preserving full dotted CLI names and CLI > YAML precedence.
- Keep native request capture, per-request version identity, direct weight transfer, full-weight/LoRA behavior, colocated memory handoff and durable recovery.
- Reuse Reef's existing executors, scenario commits, artifact records and deployment supervision.
- Make shared coordination importable without concrete model frameworks.

Non-goals:

- Implementing vLLM or MLX adapters, or declaring arbitrary backend combinations compatible.
- Cloud provisioning, model-fit planning, dynamic parallelism resizing or a general cluster autoscaler.
- Managing PRM or user-simulation services. OpenClawRL deploys those independently and consumes their endpoints through recipe configuration.
- Sending weight tensors through Reef's HTTP service or control coordinator.

### Proposed architecture

`reef/runtime/` has exactly these five modules and the generic executor package:

```text
reef/
├── runtime/
│   ├── interfaces.py
│   ├── scheduler.py
│   ├── deployment.py
│   ├── publication.py
│   ├── recovery.py
│   └── executor/
├── inference/
│   ├── runtime.py
│   ├── http.py
│   └── sglang/runtime.py
├── train/
│   ├── runtime.py
│   └── slime_backend/runtime.py
└── scenario/committer.py
```

| Module | Responsibility |
| --- | --- |
| `runtime/interfaces.py` | Independent `TrainingRuntime` / `InferenceRuntime`, native `TrainingBackend` / `InferenceBackend`, shared values, admission and storage contracts |
| `runtime/scheduler.py` | Scenario commit/admission scheduling, serialized training jobs, staleness checks and resource handoffs |
| `runtime/deployment.py` | Deployment plans, resource/component lifecycle, connection configuration and lazy runtime factory discovery |
| `runtime/publication.py` | Weight transfer ordering and verification, version allocation, commit barriers, LoRA residency and transfer locking |
| `runtime/recovery.py` | Atomic job/history persistence, startup reconstruction, inference pause/recovery and engine health monitoring |
| `runtime/executor/` | Backend-neutral workers, control RPCs, futures and coordinator discovery |
| `inference/runtime.py`, `http.py`, `model_config.py` | Generic inference runtime connection, HTTP proxy/request handling and scenario model overrides |
| `inference/sglang/` | `SGLangInferenceRuntime`, native configuration, engine/router launch, request capture, plugin, weight receipt and memory controls |
| `train/runtime.py`, `train/slime_backend/` | Generic training runtime connection; `SlimeTrainingRuntime`, native optimizer/checkpoint operations, export and sending |
| `service/runtime.py` | Generic training/inference pair assembly for existing executor/Ray connection kinds |
| `scenario/committer.py` | Existing durable artifact commits and ordering |

`runtime` is a PEP 420 namespace package: it has no package facade or `__init__.py`. Directory and import-dependency tests enforce this layout, and wheel CI verifies the same files are shipped. Extensions import contracts directly from `reef.runtime.interfaces`; factories come from `reef.runtime.deployment`. The removed `adapters/`, `control/`, `weights/`, and `training_job/` paths have no compatibility shims.

Within runtime, `interfaces` is the lowest layer; publication consumes it, recovery consumes publication, scheduler consumes recovery/publication, and deployment assembles them. Executor code depends only on interfaces and its own execution modules. Shared `TrainingJobStore` and `ScenarioHistoryStore` ABCs let publication consume durable state without importing the recovery implementation. One `FileTrainingJobStore` instance is shared by execution and publication; writes still become durable before in-memory marker transitions.

`RuntimeCandidateBackend` maps recipe batches and results and delegates coordination to `RuntimeScheduler`. The latter receives separate `TrainingRuntime` and `InferenceRuntime` objects. Recipes hold `training_runtime=training` and `runtime=inference`; runtime factories return the pair. This preserves the existing serving and candidate contracts without making training inherit inference.

The worker-side `TrainingCoordinator` is a Reef class created through `Executor`. It composes `TrainingExecution`, `TrainingPublication`, `BackendWeightPublisher` and `TrainingRecovery`, and consumes two explicit operation contracts:

- `TrainingBackend`: preparation, optimizer work, checkpoint I/O, native weight/adapter sending and training-worker lifecycle.
- `InferenceBackend`: endpoint/version reads, initial version installation, generation pause/resume, recovery, memory operations and adapter removal.

These contracts live in `reef/runtime/interfaces.py`, independently of the coordinator implementation. Native trainers such as Slime implement `TrainingBackend`. The existing `reef.train.backend.CandidateBackend` is a separate recipe-facing candidate lifecycle; `RuntimeCandidateBackend` connects it to the scheduler. Backend adapters import the operation contracts without depending on the coordinator implementation.

The native interfaces are abstract base classes. `SlimeTrainingBackend(TrainingBackend)` and `SGLangInferenceBackend(InferenceBackend)` explicitly implement them; the Slime trainer also implements the executor failure-listener interface. Managed services, resource owners, job/publication providers and engine-control adapters follow the same explicit-inheritance model. Inference factory discovery requires an `InferenceService` instance before allocation. CI forbids `typing.Protocol`, `typing_extensions.Protocol` and `runtime_checkable` across first-party Python. The existing `TYPE_CHECKING` ban continues across `reef/`, `recipes/` and `tests/`; neither ban permits baseline exceptions. Optional capabilities use separate ABCs. Existing documented Callable debt cannot grow.

The recipe candidate lifecycle is `CandidateBackend` / `RuntimeCandidateBackend`; buffered or streamed requests use `InferenceHandler`; standalone engine command descriptors use `InferenceCommand`. These names do not alias native backend contracts.

`RuntimeScheduler` and `TrainingCoordinator` remain distinct objects in `scheduler.py`: one coordinates scenario decisions and durable commit acknowledgement, while the other serializes model-worker operations. `BackendWeightPublisher` in `publication.py` performs the actual version allocation, transfer ordering, receiver verification and adapter replay. `TrainingRecovery` in `recovery.py` reconstructs startup state behind the same serving barrier.

The coordinator owns the ordering and publication policy. Native senders must not pause, resume, offload or restart inference. Native receiver operations must not decide that a successful update authorizes serving.

### Component assembly and resource ownership

`TrainingDeployment.create_training_plan(config, *, loss_family)` returns an unallocated `TrainingDeploymentPlan`: training component, coordinated resources, plain launch data, coordinator options and health policy. It does not construct an inference component.

Reef independently resolves `inference.backend` through `inference_service_for`. Built-ins load lazily; external definitions can use a `module:factory` reference or the `reef.inference_backends` entry-point group. Each inference factory constructs and validates its native configuration. The assembler creates `ModelDeploymentPlan` and validates the selected sender/receiver protocol before resource allocation.

Startup order:

1. Acquire the coordinated resource allocation.
2. Start and verify inference. Fence the receiver and release initial inference allocations before colocated trainers initialize.
3. Start training with only its supplied resources.
4. Attach a `WeightTransferSession` containing the protocol, a borrowed receiver executor and a fresh attachment identity.
5. Create Reef's coordinator from the two backend operation adapters, reconstruct durable state and verify readiness.

Shutdown retires active coordinator operations, training, inference and reservations in reverse order. All cleanup handles partial startup and repeated calls. The process owning an operational training group also observes its current workers, including deliberate worker replacement; stale driver-side monitors must not treat intentional release as a deployment failure.

Reef makes the GPU reservation itself (`reef.runtime.executor.placement`): one placement group sized from the inference capacity, the trainer's GPU count and `training.colocate`, with bundles ordered by node and device. The Slime integration receives the training and inference portions and maps them onto its actor, critic and rollout groups; independent SGLang consumes the inference portion. Colocated training and inference share the same bundles rather than allocating the same GPUs twice. Slime no longer creates placement groups, and no Slime placement code is a dependency of Reef's scheduler.

The existing `slime-sglang-control-v2` identifier describes the current native attachment protocol. It is retained for wire compatibility and is not a universal tensor format. An OpenAI-compatible URL alone does not supply native token capture, weight receipt, version installation or memory controls. Adding another backend requires an actual compatible transport and contract tests.

### Weight synchronization and publication

Reef assigns the serving identity and persists a target runtime load ID before transferring a checkpoint. Sender preparation runs while colocated inference is released: full-disk export can finish and retire training workers before receiver memory is restored. Sending then consumes the prepared target directly. A failed receive can replay retained disk output without recreating the trainer; a later export reconstructs workers only after Reef releases receiver memory. The native sender transfers to receiver workers directly. Reef checks that the sender and every receiver report the target; a partial, empty or mismatched result leaves inference fenced.

The publication lifecycle remains:

`RUNNING -> CHECKPOINT -> UPDATING_WEIGHTS -> READY_TO_COMMIT -> HEAD_COMMITTED -> COMPLETE`

`RUNNING` is durably written before optimizer execution. A checkpoint is recorded only after required model, optimizer and recovery state is persisted. `UPDATING_WEIGHTS` records an attempt whose receiver state may be mixed. A retry or cold reconstruction reuses a persisted transfer target; unchanged-weight republication preserves the committed identity.

Successful transfer does not reopen serving. The matching scenario/artifact commit must be durable before acknowledgement transitions through `HEAD_COMMITTED` and generation resumes. Failure during restoration, version checks or acknowledgement preserves the barrier. An ambiguous `RUNNING` job is never automatically retrained.

Runtime IDs keep their `<incarnation>:<sequence>` representation. The marker gains an optional `target_runtime_load_id` while a transfer is pending; older markers without it remain readable. Public artifacts, receipt linkage and provider request bodies retain their contracts.

### LoRA and colocated execution

Reef owns per-scenario adapter history and inference adapter residency. Training retains scenario slot restoration and native adapter sending; inference retains native installation and removal. Residency policy and committed names do not live inside Slime.

For colocation, Reef pauses inference and releases the required memory before optimizer work. Publication restores receiver weights and KV/graph state, transfers and verifies weights/adapters, then awaits commit acknowledgement before resuming generation. Acknowledged memory transitions are idempotent; uncertain transitions block reuse until receiver recovery.

`training.options.keep-lora-base-resident` retains only the frozen inference base during later colocated LoRA steps. Initial trainer construction still releases all inference memory. Cold recovery restores the base and every committed scenario adapter before restoring the active/pending publication. The selected workload must still fit the actual hardware.

The same managed ownership path is implemented for full-weight, LoRA, colocated and combined modes. CPU contracts demonstrate ordering and failure behavior; actual GPU checkpoint loading, IPC/NCCL transfer, memory capacity and performance remain acceptance gates.

### Recovery and observability

Reef's managed supervisor withdraws readiness, retires owned processes and native children, confirms cleanup and rebuilds both model components from durable state. The existing HTTP service and external Ray cluster remain alive. Name-based coordinator discovery reconnects existing runtime objects; discovered inference endpoints refresh, while explicitly configured gateway URLs remain fixed. Submitted mutating calls and in-flight inference requests are not automatically replayed.

Recovery allows three replacements within five minutes, with interruptible 1/2/4-second delays. Unconfirmed retirement, invalid checkpoint state or failed replacement stops the driver. `RUNNING` requires operator recovery. Slime also refuses cold reconstruction from `REJECTING`/`REJECTED` until the committed checkpoint is restored, since the newest trainer checkpoint contains the declined candidate.

Health probes remain nonblocking and do not accumulate behind active training. Replayable publication failures are handled by the publication transaction; health alone cannot reopen serving. Borrowed external engines and explicit legacy process entrypoints retain their documented ownership and do not opt into automatic model reconstruction.

### Public configuration and migration

- `inference.num-gpus`: total positive inference GPU capacity, defaulting to `inference.tensor-parallel-size`.
- `inference.tensor-parallel-size`: GPUs per tensor-parallel engine, default 1. Capacity must be compatible with the selected topology.
- `inference.options`: receiver-native options without a training-specific prefix.
- `training.backend`, `training.options`: trainer selection and native trainer options. Equivalent inference resource flags such as `rollout-num-gpus` and `rollout-num-gpus-per-engine` are rejected in managed configurations.
- `training.colocate` (default false): whether training and inference share the same GPUs. Reef sizes the reservation from it and passes the native colocation/offload flags to the driver; `training.options.colocate` and `training.options.offload-rollout` are rejected.
- Preserve CLI > YAML and complete dotted option names. Do not restore public `service`/`services` sections.
- Native backends use the paired `TrainingBackend` / `InferenceBackend` ABCs; scheduling uses the independent `TrainingRuntime` / `InferenceRuntime` ABCs. Concrete implementations inherit their interfaces.
- Python recipe extensions migrate to `CandidateBackend` and `Trainer.candidate_backend`; request extensions migrate to `InferenceHandler` and `inference_handler`. The old Slime candidate alias is removed; `SlimeTrainingBackend` names the native trainer.
- Custom request adapters use `inference.handler-factory` / `inference.handler-config`; previous request-adapter `backend-factory` / `backend-config` fields are rejected. This does not change native backend selection or entry-point group names.
- SGLang implementation paths move to `reef.inference.sglang`; examples, plugins and packaging must follow. Native configuration construction belongs there, while legacy Slime/group-file input conversion may remain a plain-data adapter at the input boundary.
- `ModelRuntime`, `ExecutorModelRuntime` and the aggregate `RayRuntime` alias are removed. Extensions implement separate runtime contracts. Generic `executor_training` and `ray_training` kinds remain; automatic Slime assembly selects `slime_training` with an independently selected `inference_runtime` (currently `sglang`). SGLang owns its default request handler. Coordinator transports are named `CoordinatorClient` and Executor/Ray variants under `runtime/executor/connection.py`.
- Direct `start_bridge`/`TrainBridgeActorImpl` construction is removed. Use component assembly and Reef's generic coordinator. The legacy Slime CLI keeps argument-file and healthcheck syntax through that assembly.
- Third-party model drivers migrate from `create_model_plan` to `create_training_plan` plus an independently selected inference definition. In-process integrations can keep their own entrypoint.

### Security, privacy and trust boundaries

Borrowed inference engines, independently deployed recipe services and external clusters are never terminated as owned model components. Keep credentials in existing environment/secret paths and out of diagnostic output. Preserve provider request bodies, scenario isolation, receipt linkage and artifact publication/recovery behavior. This proposal adds no new public HTTP control surface.

### Testing and acceptance

- [ ] Maintainers accept this RFC.
- [ ] CPU contracts verify independent backend selection, incompatible-pair rejection before allocation, cleanup and real process attachment.
- [ ] Dependency tests forbid concrete training/inference imports from runtime and cross-imports between backend integrations, including function-local imports.
- [ ] Failure tests cover partial transfers, persisted version targets, commit gating, adapter restoration, intentional worker replacement and ambiguous jobs.
- [ ] Full tests, typing, repository checks, documentation and installed-wheel imports pass for the submitted revision.
- [ ] Real GPU full-weight and LoRA checkpoint/transport acceptance passes.
- [ ] Colocated and combined-mode GPU memory/correctness/performance acceptance passes.
- [ ] Backend-extension and deployment documentation describe the implemented boundaries and unsupported combinations.

Actual command results and commit identifiers are recorded in PR #432. CPU success does not satisfy the GPU gates, and this RFC stays open until acceptance is complete.

### Alternatives considered

- Keeping Slime as the lifecycle owner couples every future backend to its rollout manager.
- Moving classes without moving coordination leaves backend-specific publication and recovery policy.
- A shared aggregate model runtime hides the training/inference boundary again.
- A universal tensor format or cluster scheduler adds unrelated scope; start with explicit supported transport pairs and existing executors.

### Related work and implementation ownership

Implementation: #432. Configuration and automatic assembly: #414/#424. Capture and future vLLM integration: #381. MLX integration: #325. Future priority/admission policy: #28. Application component overrides: #20.

Service/runtime and backend maintainers review their contracts and integrations. GPU execution and maintainer acceptance remain required before declaring completion.

### Submission checks

- [x] Existing issues, pull requests and RFCs were searched for overlapping work.
- [x] This issue is the RFC; discussion and implementation are not acceptance.
- [x] Material design changes and implementation links are maintained here.

### AI assistance

OpenAI Codex inspected the implementation and related issues and assisted with this RFC, code, tests and documentation at the contributor's request. The human contributor remains responsible for reviewing and understanding the result.
